### PR TITLE
Monit 28785 exponential histogram

### PR DESCRIPTION
**Pull request by deosu**

### CODE REVIEW

**Review by keep94 — COMMENTED** [review]
Nice, compact static void main method.

File: java-metric-example/src/main/java/com/vmware/OtelService.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import io.opentelemetry.sdk.metrics.InstrumentSelector;
+import io.opentelemetry.sdk.metrics.InstrumentType;
+import io.opentelemetry.sdk.metrics.SdkMeterProvider;
+import io.opentelemetry.sdk.metrics.SdkMeterProviderBuilder;
+import io.opentelemetry.sdk.metrics.View;
+import io.opentelemetry.sdk.metrics.export.AggregationTemporalitySelector;
+import io.opentelemetry.sdk.metrics.export.MetricReader;
+import io.opentelemetry.sdk.metrics.export.PeriodicMetricReader;
+import io.opentelemetry.sdk.metrics.internal.view.ExplicitBucketHistogramAggregation;
+import io.opentelemetry.sdk.metrics.internal.view.ExponentialHistogramAggregation;
```
Comment (oppegard) [inline java-metric-example/src/main/java/com/vmware/OtelService.java:22]: I'm concerned that we're importing internal packages – it makes me think they're not meant for users of the SDK, no? I see this [PR](https://github.com/open-telemetry/opentelemetry-java/pull/4472) was opened two days ago, allowing to configure a `MetricReader` and `MetricExporter` to use exponential histograms by default. Perhaps that's the non-internal API to use? We'd need to wait for the PR to be merged of course, before we can consume that API.

**Review by oppegard — COMMENTED** [review]
See my inline comment about use of internal packages.

File: java-metric-example/pom.xml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                                    <mainClass>com.vmware.App</mainClass>
+                                </transformer>
+                            </transformers>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
[the remark this answers is not in the document]
Comment (deosu) [reply java-metric-example/pom.xml:118]: Adding

File: java-metric-example/pom.xml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        <filters>
+            <filter>src/main/resources/env/config.${env}.properties</filter>
+        </filters>
+
+        <resources>
+            <resource>
+                <directory>src/main/resources</directory>
+                <filtering>true</filtering>
+                <includes>
+                    <include>*.properties</include>
```
[the remark this answers is not in the document]
Comment (deosu) [reply java-metric-example/pom.xml:89]: It specifies to turn on filtering by setting build.resources.resource.filtering to true. It specifies the resource directory via the build.resources.resource.directory.

File: java-metric-example/src/main/resources/config.properties [hunk 4]
```diff
@@ -0,0 +1,14 @@
+application=${application.name}
```
[the remark this answers is not in the document]
Comment (deosu) [reply java-metric-example/src/main/resources/config.properties:1]: It should be application=${application}. Thank you for pointing it out.

File: java-metric-example/src/main/java/com/vmware/App.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * @author Sumit Deo (deosu@vmware.com)
+ */
+public class App {
+  private final static Logger logger = LoggerFactory.getLogger(App.class);
+  public static void main(String[] args) throws InterruptedException {
+
+    OtelService service = new OtelService();
+    LongHistogram recorder = service.initHistogramMetricRecorder();
```
[the remark this answers is not in the document]
Comment (deosu) [reply java-metric-example/src/main/java/com/vmware/App.java:16]: We can eventually, but as of now it is sending different types of Histograms only.

File: java-metric-example/src/main/java/com/vmware/OtelService.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+import io.opentelemetry.sdk.resources.Resource;
+
+/**
+ * @author Sumit Deo (deosu@vmware.com)
+ */
+public class OtelService {
```
[the remark this answers is not in the document]
Comment (deosu) [reply java-metric-example/src/main/java/com/vmware/OtelService.java:28]: If we want to add Gauge/Sum metric in this example then we will have a multiple methods here creating different types of interfaces(LongHistogram/LongCounter, etc)

File: java-metric-example/src/main/java/com/vmware/OtelService.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    AggregationTemporalitySelector temporalitySelector;
+    if (cache.getProperty("aggregation.temporality").equals("delta")) {
+      temporalitySelector = AggregationTemporalitySelector.deltaPreferred();
+      logger.info("Delta aggregation temporality selected.");
+    }
+    else {
+      temporalitySelector = AggregationTemporalitySelector.alwaysCumulative();
+      logger.info("Cumulative aggregation temporality selected.");
+    }
+
```
[the remark this answers is not in the document]
Comment (deosu) [reply java-metric-example/src/main/java/com/vmware/OtelService.java:56]: Done

File: java-metric-example/src/main/java/com/vmware/OtelService.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    logger.info("Initializing OtlpGrpcMetricExporter.");
+    OtlpGrpcMetricExporter metricExporter = OtlpGrpcMetricExporter.builder()
+        .setEndpoint(cache.getProperty("otel.exporter.otlp.endpoint"))
+        .setAggregationTemporalitySelector(temporalitySelector)
+        .build();
+
+    logger.info("Initializing PeriodicMetricReader.");
+    MetricReader periodicReader =
+        PeriodicMetricReader.builder(metricExporter).setInterval(Duration.ofMillis(1000)).build();
```
[the remark this answers is not in the document]
Comment (deosu) [reply java-metric-example/src/main/java/com/vmware/OtelService.java:65]: Done

File: java-metric-example/src/main/java/com/vmware/OtelService.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        boundaries.add(Double.valueOf(str));
+      }
+
+      view = View.builder()
+          .setAggregation(ExplicitBucketHistogramAggregation.create(boundaries))
+          .build();
+      logger.info("Initializing View with aggregation ExplicitBucketHistogramAggregation. " +
+          "ExplicitBucketCounts: " + boundaries.size());
+    }
+
```
[the remark this answers is not in the document]
Comment (deosu) [reply java-metric-example/src/main/java/com/vmware/OtelService.java:98]: Done

File: java-metric-example/src/main/java/com/vmware/OtelService.java (hunk 1 above)
Comment (deosu) [reply java-metric-example/src/main/java/com/vmware/OtelService.java:22]: Lets hold on to this PR, we will wait for Jack's PR to be merged.

**Review by keep94 — APPROVED** [review]
LGTM

**Pull request merged**